### PR TITLE
Add ci flag

### PR DESCRIPTION
**Pull request by maxnz**
Implements #110 

- If the ci flag is present:
  - Check that course and student flags are present
  - Automatically download correct specs
  - Override highlight_partials, no_progress, no_update and no_check as true
  - Find all submitted homework directories and add these to to_record
  - Record
  - Print table
  - Check if build should pass or fail:
    - Check if any homeworks had missing files
      - If so, output which files are missing
    - Check if any files failed to compile
      - If so, output which files failed and why
    - If either of the above were true, fail the build

Also:
- Add --student as an alias for --students
- If a course is specified but no specs are downloaded (and ci flag is not present), it will only ask if it should download and then infer where it should download from based on the value of course
- Add ADS specs url
- Remove limitation on --course values to allow stogit group name to be used as the argument and parsed to find the course

I've been testing this in gitlab-ci with this .gitlab-ci.yml:
```
test:
  before_script:
    - apk update && apk add gcc g++ make python3 git
    - mkdir -p ./students/"$CI_PROJECT_NAME"
    - cp -r $(ls | grep 'hw') $(ls | grep 'lab') $(ls | grep 'ws') ./students/"$CI_PROJECT_NAME"
  script:
    - git clone https://github.com/StoDevX/cs251-toolkit.git
    - cd cs251-toolkit
    - git checkout add-ci-flag
    - python3 setup.py develop
    - cd ..
    - cs251tk --ci --course $CI_PROJECT_NAMESPACE --student $CI_PROJECT_NAME
```

The actual .gitlab-ci.yml for the students could look something like this:
```
build:
  before_script:
    - apk update && apk add gcc g++ make python3 git
    - pip3 install cs251tk
    - mkdir -p ./students/"$CI_PROJECT_NAME"
    - cp -r $(ls | grep 'hw') $(ls | grep 'lab') $(ls | grep 'ws') ./students/"$CI_PROJECT_NAME"
  script:
    - cs251tk --ci --course $CI_PROJECT_NAMESPACE --student $CI_PROJECT_NAME
```

### CODE REVIEW

**Comment by hawkrives** [comment]
You can also just use the stodevx/cs251toolkit docker image, instead of installing everything on every run. 

I'll review this soon. Thanks for working on it!

File: cs251tk/student/markdownify/process_file.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 from .pipe import pipe
 
 
-def get_file(filename, results, options):
+def get_file(filename, results, options, ci):
```
Comment (hawkrives) [inline cs251tk/student/markdownify/process_file.py]: Do we need this argument?

File: cs251tk/toolkit/__main__.py [hunk 2]
```diff
@@ -76,25 +95,37 @@ def main():
...
-                print("Class not recognized", file=sys.stderr)
+                print("data directory not found and no course specified")
                 sys.exit(1)
+
         else:
-            sys.exit(1)
+            print('data directory not found', file=sys.stderr)
+            if args['course']:
+                download = input("Download specs for {}? (Y/N)".format(args['course'].upper()))
+                if download and download.lower()[0] == "y":
```
Comment (hawkrives) [inline cs251tk/toolkit/__main__.py]: I think you can move this  line and the input line out of the conditional

File: cs251tk/toolkit/args.py [hunk 3]
```diff
@@ -28,13 +28,15 @@ def build_argparser():
     parser.add_argument('--skip-update-check', action='store_true',
                         default=getenv('CS251TK_SKIP_UPDATE_CHECK', False) is not False,
                         help='skips the pypi update check')
+    parser.add_argument('--ci', action='store_true',
+                        help='Configure for gitlab-ci usage')
 
     specs = parser.add_argument_group('control the homework specs')
-    specs.add_argument('--course', default='sd', choices=['sd', 'hd', 'ads'],
+    specs.add_argument('--course', default='sd',
```
Comment (hawkrives) [inline cs251tk/toolkit/args.py]: I think it's still be nice to list the allowed courses? The choices list is printed when you give --help.

File: cs251tk/toolkit/args.py [hunk 4]
```diff
@@ -159,6 +161,19 @@ def process_args():
...
 
+    if args['ci']:
+        if not args['course'] or not args['students']:
+            print("ci flag must be accompanied by course and student flags", file=sys.stderr)
+            sys.exit(1)
+        args['highlight_partials'] = True
+        args['no_progress'] = True
+        args['no_update'] = True
+        args['no_check'] = True
+        _, string, _ = run(['ls', 'students/{}/'.format(str(args['students'][0]).replace("['", "").replace("']", ""))])
```
Comment (hawkrives) [inline cs251tk/toolkit/args.py]: Can we replace this with some calls to glob?

**Comment by maxnz** [comment]
What would the syntax for that be?

File: cs251tk/toolkit/args.py [hunk 5]
```diff
@@ -159,6 +161,19 @@ def process_args():
...
+        for line in string.split('\n'):
+            if "hw" in line or "lab" in line or "ws" in line:
+                args['to_record'] = args['to_record'] + [[line]]
```
Comment (hawkrives) [inline cs251tk/toolkit/args.py]: Let's just call `args['to_record'].append(line)` here

**Comment by hawkrives** [comment]
@maxnz 

```
image: 'stodevx/cs251-toolkit:v2.4.3'
build:
  stage: build
  before_script: ... 
```

The image key just takes any docker image on docker hub. (I also forget that those get automatically published, but I figure we might as well take advantage of using the image that has python, the toolkit, make, g++, and other stuff installed, instead of installing it every time.)

**Comment by maxnz** [comment]
Does that work for pull request versions too?

File: cs251tk/toolkit/args.py (hunk 3 above)
Comment (maxnz) [reply cs251tk/toolkit/args.py]: The problem is that it throws an exception if I try to use `--course sd-s19`, for example. If we can somehow allow those too, then that would work.

File: cs251tk/toolkit/args.py (hunk 5 above)
Comment (maxnz) [reply cs251tk/toolkit/args.py]: It'll have to be `args['to_record'].append([line]), but sure.

File: cs251tk/toolkit/__main__.py (hunk 2 above)
Comment (maxnz) [reply cs251tk/toolkit/__main__.py]: The difference is that it has a slightly different prompt if `course` is not defined. (R118)

File: cs251tk/student/markdownify/process_file.py (hunk 1 above)
Comment (maxnz) [reply cs251tk/student/markdownify/process_file.py]: Apparently not

**Comment by hawkrives** [comment]
Not currently, but we could auto-publish docker tags for branches in the repo too.

File: cs251tk/toolkit/args.py (hunk 3 above)
Comment (hawkrives) [reply cs251tk/toolkit/args.py]: Ah. How about we put the three into the help string, then, along with an example of sd-s19?

File: cs251tk/toolkit/__main__.py (hunk 2 above)
Comment (hawkrives) [reply cs251tk/toolkit/__main__.py]: Ah. Ok.

File: cs251tk/toolkit/__main__.py [hunk 6]
```diff
@@ -45,9 +45,28 @@ def run_server(basedir):
...
+    course = course.split("-")[0].lower()
+    try:
+        url = spec_urls[course]
+    except KeyError:
+        print("Course {} not recognized".format(course))
+        sys.exit(1)
+    with chdir(basedir):
+        run(['git', 'clone', url, 'data'])
+        if not stogit:
+            return compute_stogit_url(course="hd", stogit=None, _now=datetime.date.today())
```
Comment (hawkrives) [inline cs251tk/toolkit/__main__.py]: Is this supposed to be hard-coded to Hardware?

File: cs251tk/toolkit/args.py [hunk 7]
```diff
@@ -169,10 +171,9 @@ def process_args():
         args['no_progress'] = True
         args['no_update'] = True
         args['no_check'] = True
-        _, string, _ = run(['ls', 'students/{}/'.format(str(args['students'][0]).replace("['", "").replace("']", ""))])
-        for line in string.split('\n'):
-            if "hw" in line or "lab" in line or "ws" in line:
-                args['to_record'].append([line])
+        dirs = glob('students/*/hw*') + glob('students/*/lab*') + glob('students/*/ws*')
```
[the remark this answers is not in the document]
Comment (maxnz) [reply cs251tk/toolkit/args.py]: It doesn't. It returns an empty set.

File: cs251tk/toolkit/__main__.py (hunk 6 above)
Comment (maxnz) [reply cs251tk/toolkit/__main__.py]: Nope. That's left over from copying the last version of it.

File: cs251tk/toolkit/args.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (hawkrives) [reply cs251tk/toolkit/args.py]: Drat.

**Comment by maxnz** [comment]
Will do.

**Comment by maxnz** [comment]
@Jedmeyer Here is the .gitlab-ci.yml file that uses the new update:
```
image: 'stodevx/cs251-toolkit:v2.6.0'
build:
  stage: build
  before_script:
    - mkdir -p ./students/"$CI_PROJECT_NAME"
    - cp -r $(ls | grep 'hw') $(ls | grep 'lab') $(ls | grep 'ws') ./students/"$CI_PROJECT_NAME"
  script:
    - cs251tk --ci --course $CI_PROJECT_NAMESPACE --student $CI_PROJECT_NAME
```

**Comment by hawkrives** [comment]
@maxnz Turns out that any branch pushed to this repo that starts with `travis-` will be pushed to Docker Hub under that name – so `origin/travis-ci-flag` would be tagged and pushed as `stodevx/cs251-toolkit:travis-ci-flag`. If you find yourself needing/wanting that.

**Comment by maxnz** [comment]
Good to know. Thanks

**Pull request merged**